Dev

Cerfoglg opened the pull request:
Fixed issues: #19, #18, #17, #16

Closes #18, #19, #17, #16, #11

Cerfoglg committed:
Removed .settings

Cerfoglg committed:
Changed "cpu" to "stats"

Cerfoglg committed:
Fixed names

Cerfoglg committed:
Switched to gzip for all collectors

Cerfoglg committed:
Added CSV format saving for a table

Cerfoglg committed:
Changed Minio key naming

VincenzoFerme commented on dbms/sqlquery/mysql/mysqldump/src/cloud/benchflow/collectors/Dump.go:
```diff
@@ -7,9 +7,25 @@ import (
+    "strings"
 )
 
+var runCounter int
```
Why do you have this variable?

Cerfoglg replied:
It was mostly for testing, since we didn't have a defined Minio structure at the time.

VincenzoFerme replied:
Ok. Move it in and ENV variable.

VincenzoFerme commented on dbms/sqlquery/mysql/mysqldump/src/cloud/benchflow/collectors/Dump.go:
```diff
@@ -7,9 +7,25 @@ import (
 )
 
+var runCounter int
+
```
Refactor it in different functions, avoiding duplicated code.

Cerfoglg committed:
Removed run counter for now

Cerfoglg committed:
Added channels for stopping, cleaned code

Cerfoglg commented:
Worked on issue #11

There is still an issue with stopping the actual collection. In the API there is a "Done" channel used in theory to signal the function receiving the stats to stop, however it's not well explained in the docs, and when I try sending something on that channel I get an io error that says "read/write on a closed pipe", implying the channel is closed. 

A link to the doc: https://godoc.org/github.com/fsouza/go-dockerclient#Client.Stats

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -16,6 +16,7 @@ type Container struct {
 	ID           string
 	statsChannel chan *docker.Stats
 	flagsChannel chan bool
```
Rename to doneChannel

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -16,6 +16,7 @@ type Container struct {
+	stopChannel chan bool
 }
 
 var containers [10]Container
```
Why [10]?

Cerfoglg replied:
Need to fix that so it's dynamic

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -110,34 +131,16 @@ func stopCollecting(w http.ResponseWriter, r *http.Request) {
+		c.stopChannel <- true
+		}
 	collecting = false
 	fmt.Fprintf(w, "Stopped collecting")
```
This is not enough, you need a waiting group to be sure about the time the goroutine effectively ends: http://adampresley.com/2015/02/16/waiting-for-goroutines-to-finish-running-before-exiting.html#waiting-for-them-to-finish

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -41,64 +42,84 @@ func monitorStats(container *Container) {
 		fmt.Fprintf(w, "Already collecting")
 		return
 	}
+	path := os.Getenv("DOCKER_CERT_PATH")
```
Why do you define the connection here and not in the main?

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -41,64 +42,84 @@ func monitorStats(container *Container) {
+		c := Container{ID: each, statsChannel: statsChannel, flagsChannel: flagsChannel, stopChannel : stopChannel}
+		containers[i] = c
+		collectStats(*client, containers[i])
 		monitorStats(&containers[i])
```
Why did you choose the name monitorStats?

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -110,34 +131,16 @@ func stopCollecting(w http.ResponseWriter, r *http.Request) {
+	for _, c := range containers {
+		c.stopChannel <- true
+		}
 	collecting = false
```
I guess you can get rid of the collecting variable if you implement the WaitGroup

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -110,34 +131,16 @@ func stopCollecting(w http.ResponseWriter, r *http.Request) {
 }
 
 func main() {
 	collecting = false
```
I guess you can get rid of the collecting variable if you implement the WaitGroup

Cerfoglg replied:
WaitGroup doesn't let you check how many routines it's waiting for, so I need to keep that collecting variables for the sake of the handlers.

VincenzoFerme replied:
ok

Cerfoglg committed:
Added WaitGroup, made containers variable dynamic, better names

Cerfoglg commented:
Added WaitGroup for synchronization, made the containers variable dynamic with slices instead of limiting its size, changed flagsChannel to doneChannel, changed monitorStats to collectStats, and previous collectStats to attachToDocker, made a separate method for connecting to the docker api and returning the client

VincenzoFerme commented:
@Cerfoglg let me know when all the issues in this pull request are fixed, so that I can work on it.

Cerfoglg commented:
@VincenzoFerme  The issues in this pull request should all be fixed now

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -0,0 +1,159 @@
+	ID           string
+	statsChannel chan *docker.Stats
+	doneChannel chan bool
+	stopChannel chan bool
```
You don't need a stopchannel for each container. One should be enough.

Cerfoglg replied:
I did it to avoid possible confusion, like having a routine stop and start immediately after, and then get stopped while leaving another hanging. Unlikely scenario, but you never know.

I'll change it to a single channel if you think that's not a problem.

VincenzoFerme replied:
This situation should not be possible because you use the collecting variable to block any new start, if there is an already running collection. The collecting variable is set to false only after all the routines have exited (using the Wait instruction).

Cerfoglg replied:
Changed to use a single stop channel.

Cerfoglg committed:
Made a single stop channel, resetting container list when restarting

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -0,0 +1,158 @@
+		fmt.Fprintf(w, "Currently not collecting")
+		return
+	}
+	for _, _ = range containers {
```
Now that you have only one channel, you can stop more goroutines by simply closing it: http://golang-examples.tumblr.com/post/64474189870/stop-goroutine-by-closing-a-quit-channel

Cerfoglg replied:
Oh right, I forgot.

VincenzoFerme commented on environment/stats/src/cloud/benchflow/collectors/stats.go:
```diff
@@ -0,0 +1,158 @@
+type Container struct {
+	ID           string
+	statsChannel chan *docker.Stats
+	doneChannel chan bool
```
Try if the same approach of the stopChannel works also for the doneChannel. It depends on how the library handle this channel.

Cerfoglg replied:
Already tried that, still gives me the same error, as if I was closing a channel that's already closed. I'll try testing this separately and see if the issue is somewhere in my code structure, but if it isn't then I'll open an issue on the docker client API github and see if they have an answer, and if it's a bug of their API they can fix it.

VincenzoFerme replied:
Sure, this is the right approach. But I guess you can also use only one channel for this.

Cerfoglg committed:
Fixed some lines
